src/webnn_native/nnapi/GraphNnapi.cpp

```diff
@@ -0,0 +1,644 @@
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "webnn_native/nnapi/GraphNnapi.h"
+
+#include <errno.h>
+#include <sys/mman.h>
+#include <unistd.h>
+#include <algorithm>
+#include <string>
+#include <vector>
+
+#include "NnapiUtils.h"
+#include "common/Assert.h"
+#include "common/Log.h"
+#include "nnapi_implementation.h"
+#include "webnn_native/ErrorData.h"
+#include "webnn_native/NamedInputs.h"
+#include "webnn_native/NamedOperands.h"
+#include "webnn_native/NamedOutputs.h"
+#include "webnn_native/Utils.h"
+
+#define WEBNN_ASSERT(condition, message) \
+    do {                                 \
+        dawn::ErrorLog() << message;     \
+        DAWN_ASSERT(condition);          \
+    } while (0)
+
+namespace webnn_native { namespace nnapi {
+
+    Graph::Graph(Context* context) : GraphBase(context) {
+        operandCount = 0;
+
+        mScalarInt32Operand.type = ANEURALNETWORKS_INT32;
+        mScalarInt32Operand.dimensionCount = 0;
+        mScalarInt32Operand.dimensions = nullptr;
+        mScalarInt32Operand.scale = 0.0f;
+        mScalarInt32Operand.zeroPoint = 0;
+
+        mScalarBoolOperand.type = ANEURALNETWORKS_BOOL;
+        mScalarBoolOperand.dimensionCount = 0;
+        mScalarBoolOperand.dimensions = nullptr;
+        mScalarBoolOperand.scale = 0.0f;
+        mScalarBoolOperand.zeroPoint = 0;
+
+        mNnapiMgr = new NnapiManager();
+    }
+
+    Graph::~Graph() {
+        for (auto node : mGraphOperandInfo) {
+            if (node.second.mem)
+                mNnapiMgr->FreeMemory(node.second.mem);
+
+            close(node.second.fd);
+        }
+
+        delete mNnapiMgr;
+    }
+
+    MaybeError Graph::AddConstant(const op::Constant* constant) {
+        NodeInfo node;
+        void* buffer = const_cast<void*>(constant->GetBuffer());
+
+        CreateNodeFromOperandDescriptor(constant->GetOperandDescriptor(), &node);
+        DAWN_TRY(mNnapiMgr->CreateOperandAndSetMemory("const", &node, buffer));
+
+        mGraphOperandInfo[node.opIndex] = node;
+        mGraphNodeMap[constant->PrimaryOutput()] = node.opIndex;
+
+        return {};
+    }
+
+    MaybeError Graph::AddInput(const op::Input* input) {
+        NodeInfo node;
+
+        CreateNodeFromOperandDescriptor(input->GetOperandDescriptor(), &node);
+        DAWN_TRY(mNnapiMgr->CreateInputOutputOperand(input->GetName(), &node));
+
+        mGraphOperandInfo[node.opIndex] = node;
+        mGraphNodeMap[input->PrimaryOutput()] = node.opIndex;
```

Comment:
You need to get the index first with `mGraphNodeMap`, then get the `NodeInfo` from `mGraphOperandInfo`, why don't you set the nodeInfo to `mGraphNodeMap`, and remove the `mGraphOperandInfo` member variable?